### PR TITLE
#2266 Custom port images

### PR DESCRIPTION
**Pull request by tomlyn**
Fixes: #2266 
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: canvas_modules/common-canvas/src/object-model/layout-dimensions.js [hunk 1]
```diff
@@ -211,18 +233,28 @@ const portsHorizontalDefaultLayout = {
 		// this: { x_pos: 5, y_pos: 10, pos: "topRight" }. x_pos and y_pos are
 		// offsets from the pos point on the node.
 		// The order of the elements corresponds to the order of ports in the
-		// outputs array for the node.
+		// outputs array for the node. If there are more input ports than elements
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/object-model/layout-dimensions.js:236]: typo, should be `output`

File: canvas_modules/common-canvas/src/object-model/layout-dimensions.js [hunk 2]
```diff
@@ -211,18 +233,28 @@ const portsHorizontalDefaultLayout = {
...
+		// Each element can have a number of different structures like this:
+		// Either
+		// { type: "circle" }
+		// Or
+		// { type: "image", src: "path/picture.svg", width: 10, height: 10 }
+		// Or
+		// { type: "jsx", src: (<FaceCool />), width: 16, height: 16 }
+		//
+		// The order of the elements corresponds to the order of ports in the
+		// inputs array for the node. If there are more input ports than elements
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/object-model/layout-dimensions.js:253]: typo, should be `output`

File: canvas_modules/common-canvas/src/object-model/layout-dimensions.js [hunk 3]
```diff
@@ -637,18 +691,28 @@ const portsVerticalDefaultLayout = {
 		// this: { x_pos: 5, y_pos: 10, pos: "topRight" }. x_pos and y_pos are
 		// offsets from the pos point on the node.
 		// The order of the elements corresponds to the order of ports in the
-		// outputs array for the node.
+		// outputs array for the node. If there are more input ports than elements
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/object-model/layout-dimensions.js:694]: typo, should be `output`

File: canvas_modules/common-canvas/src/object-model/layout-dimensions.js [hunk 4]
```diff
@@ -637,18 +691,28 @@ const portsVerticalDefaultLayout = {
...
+		// Each element can have a number of different structures like this:
+		// Either
+		// { type: "circle" }
+		// Or
+		// { type: "image", src: "path/picture.svg", width: 10, height: 10 }
+		// Or
+		// { type: "jsx", src: (<FaceCool />), width: 16, height: 16 }
+		//
+		// The order of the elements corresponds to the order of ports in the
+		// inputs array for the node. If there are more input ports than elements
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/object-model/layout-dimensions.js:711]: typo, should be `output`

File: docs/pages/03.06.01-node-customization.md [hunk 5]
```diff
@@ -207,18 +229,28 @@ The possible node layout properties are shown below with the values they have wh
...
+    // Each element can have a number of different structures like this:
+    // Either
+    // { type: "circle" }
+    // Or
+    // { type: "image", src: "path/picture.svg", width: 10, height: 10 }
+    // Or
+    // { type: "jsx", src: (<FaceCool />), width: 16, height: 16 }
+    //
+    // The order of the elements corresponds to the order of ports in the
+    // inputs array for the node. If there are more input ports than elements
```
Comment (caritaou) [inline docs/pages/03.06.01-node-customization.md:249]: typo, should be `output`

**Review by caritaou — COMMENTED** [review]
I see an error on the console in the Network example when dragging a link out
image

File: canvas_modules/common-canvas/src/object-model/layout-dimensions.js (hunk 1 above)
Comment (tomlyn) [reply canvas_modules/common-canvas/src/object-model/layout-dimensions.js:236]: Good catch - fixed

File: canvas_modules/common-canvas/src/object-model/layout-dimensions.js (hunk 2 above)
Comment (tomlyn) [reply canvas_modules/common-canvas/src/object-model/layout-dimensions.js:253]: Another good catch - fixed

File: canvas_modules/common-canvas/src/object-model/layout-dimensions.js (hunk 3 above)
Comment (tomlyn) [reply canvas_modules/common-canvas/src/object-model/layout-dimensions.js:694]: Yep good one - fixed

File: canvas_modules/common-canvas/src/object-model/layout-dimensions.js (hunk 4 above)
Comment (tomlyn) [reply canvas_modules/common-canvas/src/object-model/layout-dimensions.js:711]: Fixed

File: docs/pages/03.06.01-node-customization.md (hunk 5 above)
Comment (tomlyn) [reply docs/pages/03.06.01-node-customization.md:249]: I copied and pasted the whole text from the layout-dimensions.js file so this is fixed

**Comment by tomlyn** [comment]
I fixed the Network diagram problem -- thanks for trying that I had missed it. The code now sets default width and height for port images or guide images if they are missing from the application's node layout fields.

**Pull request merged**